### PR TITLE
Brownfield main

### PR DESCRIPTION
**Pull request by TimWeberRE100**
Completed brownfield algorithm for grid squares. Includes integration of depression_volume_finding into screening.

### CODE REVIEW

File: src/pairing.cpp [hunk 1]
```diff
@@ -148,13 +151,24 @@ Pair *check_good_pair(RoughReservoir* upper, RoughReservoir* lower,
...
   if ((max(upper->volumes) < required_volume) ||
-      (max(lower->volumes) < required_volume))
+      (max(lower->volumes) < required_volume)) {
     return NULL;
+  }
 
-  if (search_config.search_type == SearchType::PIT &&
-      !determine_pit_elevation_and_volume(upper, lower, energy_capacity,
-                                          pit_details, required_volume, head))
-    return NULL;
```
Comment (ryanstocks00) [inline src/pairing.cpp:157]: Does removing this break the single pit search?

File: src/pairing.cpp [hunk 2]
```diff
@@ -169,15 +183,17 @@ Pair *check_good_pair(RoughReservoir* upper, RoughReservoir* lower,
...
-                                            pit_details.reservoir.elevation,
+                                 pit_volume(pit_details[i],
+                                            pit_details[i].reservoir.elevation,
                                             upper->elevation),
-                             get_volumes(pit_details),
-                             int_to_double_vector(get_altitudes(pit_details))) -
+                             get_volumes(pit_details[i]),
+                             int_to_double_vector(get_altitudes(pit_details[i]))) -
           upper->elevation;
+      }
```
Comment (ryanstocks00) [inline src/pairing.cpp]: I'm a little confused, why do we iterate over all i here and keep replacing the value of upper dam wall height?

File: src/pairing.cpp [hunk 3]
```diff
@@ -169,15 +183,17 @@ Pair *check_good_pair(RoughReservoir* upper, RoughReservoir* lower,
                                              dam_wall_heights,
                                              upper->dam_volumes);
   } else {
-    if (search_config.search_type == SearchType::PIT)
-      upper_dam_wall_height =
```
Comment (ryanstocks00) [inline src/pairing.cpp:173]: Potentially same issue breaking single pit search

File: src/pairing.cpp [hunk 4]
```diff
@@ -190,29 +206,33 @@ Pair *check_good_pair(RoughReservoir* upper, RoughReservoir* lower,
...
-                                            pit_details.reservoir.elevation,
+                                 pit_volume(pit_details[i],
+                                            pit_details[i].reservoir.elevation,
                                             lower->elevation),
-                             get_volumes(pit_details),
-                             int_to_double_vector(get_altitudes(pit_details))) -
+                             get_volumes(pit_details[i]),
+                             int_to_double_vector(get_altitudes(pit_details[i]))) -
           lower->elevation;
+      }
```
Comment (ryanstocks00) [inline src/pairing.cpp]: same here

File: src/pairing.cpp [hunk 5]
```diff
@@ -337,12 +355,14 @@ void pairing(vector<unique_ptr<RoughReservoir>> &upper_reservoirs,
...
              category_cutoffs[0].power_cost) *
             (1 + tolerance_on_FOM);
+        
         if (check_good_pair(upper_reservoir, lower_reservoir,
                             tests[itest].energy_capacity,
                             tests[itest].storage_time, &temp_pair, max_FOM)) {
           temp_pairs[itest].insert(temp_pair);
+
           if ((int)temp_pairs[itest].size() > max_lowers_per_upper ||
-              (search_config.search_type == SearchType::PIT && temp_pairs[itest].size() > 1))
```
Comment (ryanstocks00) [inline src/pairing.cpp:345]: Do we need to break the single pit search?

File: src/phes_base.cpp [hunk 6]
```diff
@@ -367,12 +368,187 @@ RoughBfieldReservoir existing_reservoir_to_rough_reservoir(ExistingReservoir r){
...
-	ExistingPit pit;
+vector<ExistingPit> get_pit_details(GridSquare grid_square){
+	vector<ExistingPit> gridsquare_pits;
+
 	vector<ExistingPit> pits = read_existing_pit_data(convert_string(file_storage_location+"input/existing_reservoirs/"+existing_reservoirs_csv));
 
-	for(ExistingPit p : pits)
-		if(p.reservoir.identifier==pitname)
-			pit = p;
-	return pit;
```
Comment (ryanstocks00) [inline src/phes_base.cpp]: Can we keep this version of the function as well so that we can do single pits?

File: tools/TN_debugging.cpp [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+  }
+  printf("\n");
+}
+void print_v(deque<ArrayCoordinateWithHeight> vector) {
+  for (uint i = 0; i<vector.size(); i++){
+    printf("%d %d %.2f, ", vector[i].row, vector[i].col, vector[i].h);
+  }
+  printf("\n");
+}
+///////////////////////////////////////////
```
Comment (ryanstocks00) [inline tools/TN_debugging.cpp]: ?

**Review by ryanstocks00 — CHANGES_REQUESTED** [review]

File: src/pairing.cpp (hunk 1 above)
Comment (TimWeberRE100) [reply src/pairing.cpp:157]: Added single pit search back in

File: src/pairing.cpp (hunk 2 above)
Comment (TimWeberRE100) [reply src/pairing.cpp]: Should be for single pit, fixed now

File: src/pairing.cpp (hunk 3 above)
Comment (TimWeberRE100) [reply src/pairing.cpp:173]: Added single pit search back in

File: src/pairing.cpp (hunk 4 above)
Comment (TimWeberRE100) [reply src/pairing.cpp]: Added single pit search back in

File: src/pairing.cpp (hunk 5 above)
Comment (TimWeberRE100) [reply src/pairing.cpp:345]: Added single pit search back in

File: src/phes_base.cpp (hunk 6 above)
Comment (TimWeberRE100) [reply src/phes_base.cpp]: Added single pit search back in

File: processes [hunk 8]
```diff
@@ -1,4 +1,4 @@
-screening
-pairing
-pretty_set
-constructor
\ No newline at end of file
+screening bulk_pit
+pairing bulk_pit
+pretty_set bulk_pit
+constructor bulk_pit
```
Comment (ryanstocks00) [inline processes]: The bulk_pit should go in the task list not processes

File: tasks [hunk 9]
```diff
@@ -1,8 +1,109 @@
...
+98 8
+99 8
+100 8
+104 8
+105 8
+106 8
+111 8
+116 8
+117 8
+118 8
```
Comment (ryanstocks00) [inline tasks]: Lets not change the example tasks file

File: src/pairing.cpp [hunk 10]
```diff
@@ -148,13 +152,30 @@ Pair *check_good_pair(RoughReservoir* upper, RoughReservoir* lower,
...
   double required_volume = find_required_volume(energy_capacity, head);
   if ((max(upper->volumes) < required_volume) ||
-      (max(lower->volumes) < required_volume))
+      (max(lower->volumes) < required_volume)) {
     return NULL;
+  }
+  ExistingPit single_pit;
+
+  if (search_config.search_type == SearchType::BULK_PIT) {
+    single_pit = pit_details[0];
```
Comment (ryanstocks00) [inline src/pairing.cpp]: Do we need the `single_pit = pit_details[0]`? Seems like it could hide bugs

**Review by ryanstocks00 — CHANGES_REQUESTED** [review]

File: processes (hunk 8 above)
Comment (TimWeberRE100) [reply processes]: Changed back to default values

File: tasks (hunk 9 above)
Comment (TimWeberRE100) [reply tasks]: Changed back to default values

File: src/pairing.cpp (hunk 10 above)
Comment (TimWeberRE100) [reply src/pairing.cpp]: Removed this line and added error handling if single_pit is not assigned a value

**Review by ryanstocks00 — DISMISSED** [review]

**Review by ryanstocks00 — APPROVED** [review]

**Pull request merged**